### PR TITLE
Permutations

### PR DESCRIPTION
**Pull request by JakubValtar**
This is the first stab at permutations. Outputting k-permutations lexicographically is a quite a pain, so instead I compose it with `Combinations` and return all permutations of each k-combination.

- Is composition desirable? I haven't seen an interator in `std` which does that, though I haven't been looking that thoroughly.
- Is it ok that the order is not lexicographic?
- Is it ok that there is an extra clone per k-combination observable by the user?

### CODE REVIEW

File: src/lib.rs [hunk 1]
```diff
@@ -137,15 +244,28 @@ mod test {
     }
 
     #[test]
-    fn none_on_size_too_big() {
+     fn combinations_none_on_size_too_big() {
```
Comment (yoshuawuyts) [inline src/lib.rs]:
```suggestion
    fn combinations_none_on_size_too_big() {
```

File: src/lib.rs [hunk 2]
```diff
@@ -116,6 +130,99 @@ where
...
+    i: usize,
+    first: bool,
+}
+
+impl<T, const N: usize> FullPermutations<T, N> {
+    fn new(a: [T; N]) -> Self {
+        Self {
+            a,
+            c: [0; N],
+            i: 0,
```
Comment (yoshuawuyts) [inline src/lib.rs:145]: Probably something for a follow-up PR; but it'd be good to have more descriptive names for these variables. If we want to add this to the stdlib we should expect this code will be maintained for many years to come by people other than us.

**Review by yoshuawuyts — APPROVED** [review]
Really cool! -- one nit, but that can be fixed in a follow-up PR.

I'm going to write a docs PR next, and maybe move the two adapters into separate files so that that we can start to get things into a shape where it can be PR'd into the stdlib.

File: src/lib.rs (hunk 2 above)
Comment (JakubValtar) [reply src/lib.rs:145]: I agree, for now I kept the names the same as in the pseudocode, but we should not depend on that and we should make it clear how it works.

**Pull request merged**